pallene/checker.lua

```diff
@@ -765,19 +860,33 @@ function FunChecker:check_exp_synthesize(exp)
         local f_type = exp.exp._type
 
         if f_type._tag == "types.T.Function" then
+            local nlast_args = #exp.args
+            local last_arg = exp.args[nlast_args]
+
+            if  last_arg and (last_arg._tag == "ast.Exp.CallFunc" or
+                last_arg._tag == "ast.Exp.CallMethod")
+            then
+                self:expand_function_returns(f_type.arg_types, exp.args,
+                    last_arg)
+            end
+
+            for i = 1, #exp.args do
+                if f_type.arg_types[i] then
+                    exp.args[i] = self:check_exp_verify(exp.args[i],
+                                    f_type.arg_types[i],
+                                    "argument %d of call to function", i)
+                end
+            end
+
             if #f_type.arg_types ~= #exp.args then
                 type_error(exp.loc,
                     "function expects %d argument(s) but received %d",
                     #f_type.arg_types, #exp.args)
             end
-            for i = 1, math.min(#f_type.arg_types, #exp.args) do
-                exp.args[i] = self:check_exp_verify(
-                    exp.args[i], f_type.arg_types[i],
-                    "argument %d of call to function", i)
-            end
-            assert(#f_type.ret_types <= 1)
-            if #f_type.ret_types >= 1 then
-                exp._type = f_type.ret_types[1]
+
+            if #f_type.ret_types > 0 then
+                exp._type  = f_type.ret_types[1]
+                exp._types = f_type.ret_types
```

Comment:
I think this comment needs a clarification because it wasn't 100% accurate:

* Every function call expression should receive a `._types`, even when `#ret_types == 0`.
* There should be no need to test `if fexp._types` if we know that `fexp` is a function call expression. Every function call expression should have a `_types` now.